tests/ui/while_let_some_result.rs

```diff
@@ -0,0 +1,37 @@
+// run-rustfix
+
+#![warn(clippy::while_let_some_result)]
+
+struct Wat {
+    counter: i32,
+}
+
```

Comment:
Perhaps add a test for `foo.ok()` where `foo` is `Box<Result<..>>`? The lint should (correctly) not lint for it now though I suspect we could extend the lint such that it is able to suggest things like `if let Ok(..) = *foo`

Reply:
Apologies, I am not familiar with this `Box<Result<..>>` formatting. Would it look something like this?

```
fn base_3(foo: Box<Result<()>>) {
    while let Some(a) = foo.ok() {
        println("{}", a)
    }
}
```

Reply:
So the thing is that if `foo` is `Box<Result<T, E>>`, `foo.ok()` is `Option<T>`, _but_ `foo` does not _directly_ pattern match against `Ok()` since it's a pointer.

But yes, that's how to write the test, and the lint as currently stated should not suggest a fix (unless it can add a `*` before the `foo`

Reply:
This is still throwing me off for some reason. Apologies again, still learning Rust, and so these syntax mistakes are hard for me to follow

I've created a test case that looks like the following:

```
fn base_3(foo: Box<Result<T, E>>) {
    while let Some(a) = foo.ok() {}
}
```

The compiler does not like this test in its current state, claiming that `T` cannot be found in the current scope and throwing errors. I've tried replacing with standard types like `bool`, but still receive the same error. I've been looking around online to see if there are examples to go off of, but haven't been able to find as such

Do you have any suggested background on what this new test is trying to cover, and how I can cover it?